File: library/src/main/java/su/levenetc/android/textsurface/TextSurface.java
package su.levenetc.android.textsurface;

import android.content.Context;
import android.graphics.Canvas;
import android.util.AttributeSet;
import android.widget.FrameLayout;

import java.util.Iterator;
import java.util.LinkedList;
import java.util.TreeSet;

import su.levenetc.android.textsurface.animations.AnimationsSet;
import su.levenetc.android.textsurface.contants.TYPE;
import su.levenetc.android.textsurface.interfaces.ICameraAnimation;
import su.levenetc.android.textsurface.interfaces.ISet;
import su.levenetc.android.textsurface.interfaces.ISurfaceAnimation;
import su.levenetc.android.textsurface.interfaces.ITextSurfaceAnimation;

/**
 * Created by Eugene Levenetc.
 */
public class TextSurface extends FrameLayout {

	private TreeSet<Text> textsTree = new TreeSet<>();
	private SurfaceCamera camera = new SurfaceCamera();
	private ISurfaceAnimation currentAnimation = null;

	public TextSurface(Context context) {
		super(context);
		config();
	}

	public TextSurface(Context context, AttributeSet attrs) {
		super(context, attrs);
		config();
	}

	private void config() {
		setWillNotDraw(false);
	}

	public SurfaceCamera getCamera() {
		return camera;
	}

	public void play(TYPE type, ISurfaceAnimation... animations) {
		play(new AnimationsSet(type, animations));
	}

	public void play(ISurfaceAnimation... animations) {
		play(new AnimationsSet(TYPE.PARALLEL, animations));
	}

	public void play(ISurfaceAnimation animation) {
		configAnimations(animation);
		animation.setTextSurface(this);
		layout();
		currentAnimation = animation;
		currentAnimation.start(null);
	}

	private void layout() {
		Iterator<Text> iterator = textsTree.iterator();
		while (iterator.hasNext()) iterator.next().layout(this);
	}

	@Override protected void onLayout(boolean changed, int left, int top, int right, int bottom) {
		super.onLayout(changed, left, top, right, bottom);
	}

	private void configAnimations(ISurfaceAnimation animation) {

		if (animation instanceof ICameraAnimation) {
			((ICameraAnimation) animation).setCamera(camera);
		} else if (animation instanceof ISet) {
			LinkedList<ISurfaceAnimation> animations = ((ISet) animation).getAnimations();
			for (ISurfaceAnimation a : animations) configAnimations(a);
		} else if (animation instanceof ITextSurfaceAnimation) {
			ITextSurfaceAnimation textAnimation = (ITextSurfaceAnimation) animation;
			Text text = textAnimation.getText();
			if (text != null && textsTree.add(text)) textAnimation.setInitValues(text);
		}
	}

	@Override protected void onDraw(Canvas canvas) {
		super.onDraw(canvas);
		camera.onDraw(canvas);
		for (Text text : textsTree) text.onDraw(canvas, this);
	}

	public void reset() {
		if (currentAnimation != null) {
			currentAnimation.cancel();
			currentAnimation = null;
		}
		textsTree.clear();
		camera.reset();
		invalidate();
	}
}

File: library/src/main/java/su/levenetc/android/textsurface/TextSurface.kt
package su.levenetc.android.textsurface

import android.content.Context
import android.graphics.Canvas
import android.util.AttributeSet
import android.view.View
import su.levenetc.android.textsurface.animations.SurfaceAnimation
import su.levenetc.android.textsurface.animations.TextSurfaceAnimation
import su.levenetc.android.textsurface.animations.camera.CameraAnimation
import su.levenetc.android.textsurface.animations.sets.AnimationsSet
import su.levenetc.android.textsurface.animations.sets.Parallel
import su.levenetc.android.textsurface.animations.sets.Sequential
import su.levenetc.android.textsurface.camera.SurfaceCamera
import su.levenetc.android.textsurface.utils.getDesiredSize
import su.levenetc.android.textsurface.utils.measureDimension

/**
 * Created by Eugene Levenetc.
 */
class TextSurface : View {

    private val camera = SurfaceCamera()
    private val texts = mutableSetOf<Text>()
    private lateinit var currentAnimation: SurfaceAnimation

    init {
        setWillNotDraw(false)
    }

    constructor(context: Context?) : super(context!!)
    constructor(context: Context?, attrs: AttributeSet?) : super(context!!, attrs)

    fun playParallel(vararg animations: SurfaceAnimation) {
        play(Parallel(* animations))
    }

    fun playSequential(vararg animations: SurfaceAnimation) {
        play(Sequential(* animations))
    }

    fun play(animation: SurfaceAnimation) {
        configAnimations(animation)
        animation.textSurface = this
        layout()
        currentAnimation = animation
        currentAnimation.start()
    }

    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)
        camera.onDraw(canvas)
        texts.forEach { it.onDraw(canvas, this) }
    }

    fun reset() {
        if (this::currentAnimation.isInitialized) currentAnimation.cancel()
        texts.clear()
        camera.reset()
        invalidate()
    }

    private fun layout() {
        texts.forEach { it.layout(this) }
    }

    private fun configAnimations(animation: SurfaceAnimation) {
        if (animation is CameraAnimation) {
            animation.camera = camera
        } else if (animation is AnimationsSet) {
            animation.animations.forEach { configAnimations(it) }
        } else if (animation is TextSurfaceAnimation) {
            val text = animation.text
            if (texts.add(text)) animation.setInitValues(text)
        }
    }

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        val desiredWidth = getDesiredSize(layoutParams.width, suggestedMinimumWidth, widthMeasureSpec)
        val desiredHeight = getDesiredSize(layoutParams.height, suggestedMinimumHeight, heightMeasureSpec)

        setMeasuredDimension(
                measureDimension(desiredWidth, widthMeasureSpec),
                measureDimension(desiredHeight, heightMeasureSpec)
        )
    }
}